pallets/loans-ref/src/types/mod.rs

```diff
@@ -79,44 +77,108 @@ pub enum CloseLoanError {
-pub struct PortfolioValuation<Balance> {
-	// Computed portfolio valuation for the given pool
+#[derive(Encode, Decode, Clone, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(MaxElems))]
+pub struct PortfolioValuation<Balance, ElemId, MaxElems: Get<u32>> {
+	/// Computed portfolio valuation for the given pool
 	value: Balance,
 
-	// Last time when the portfolio valuation was calculated for the entire pool
-	last_updated: Moment,
+	/// Last time when the portfolio valuation was calculated for the entire
+	/// pool. None if never has been computed entirely.
+	last_updated: Option<Moment>,
+
+	/// Individual valuation of each element that compose the value of the
+	/// portfolio
+	values: BoundedVec<(ElemId, Balance), MaxElems>,
 }
 
-impl<Balance> PortfolioValuation<Balance>
+impl<Balance, ElemId, MaxElems> Default for PortfolioValuation<Balance, ElemId, MaxElems>
 where
-	Balance: tokens::Balance,
+	Balance: Zero,
+	MaxElems: Get<u32>,
 {
-	pub fn new(value: Balance, when: Moment) -> Self {
+	fn default() -> Self {
 		Self {
-			value,
-			last_updated: when,
+			value: Balance::zero(),
+			last_updated: None,
+			values: BoundedVec::default(),
 		}
 	}
+}
 
+impl<Balance, ElemId, MaxElems> PortfolioValuation<Balance, ElemId, MaxElems>
+where
+	Balance: EnsureAdd + EnsureSub + Ord + Copy,
+	ElemId: Eq,
+	MaxElems: Get<u32>,
+{
 	pub fn value(&self) -> Balance {
 		self.value
 	}
 
-	pub fn last_updated(&self) -> Moment {
+	pub fn last_updated(&self) -> Option<Moment> {
 		self.last_updated
 	}
 
-	pub fn update_with_pv_diff(
+	pub fn value_of(&self, id: ElemId) -> Option<&Balance> {
+		self.values
+			.iter()
+			.find(|(elem_id, _)| *elem_id == id)
+			.map(|(_, balance)| balance)
+	}
+
+	pub fn update(
 		&mut self,
-		old_pv: Balance,
-		new_pv: Balance,
-	) -> Result<(), ArithmeticError> {
-		match new_pv.cmp(&old_pv) {
-			Ordering::Greater => self.value.ensure_add_assign(new_pv.ensure_sub(old_pv)?),
-			Ordering::Less => self.value.ensure_sub_assign(old_pv.ensure_sub(new_pv)?),
-			Ordering::Equal => Ok(()),
-		}
+		pv_list: Vec<(ElemId, Balance)>,
+		when: Moment,
+	) -> Result<Balance, DispatchError> {
+		self.values = pv_list
+			.try_into()
+			.map_err(|_| DispatchError::Other("TODO"))?;
+
+		self.value = self.values.iter().try_fold(
+			Balance::zero(),
+			|sum, (_, value)| -> Result<Balance, DispatchError> { Ok(sum.ensure_add(*value)?) },
+		)?;
+
+		self.last_updated = Some(when);
+
+		Ok(self.value)
+	}
+
+	pub fn insert_elem(&mut self, id: ElemId, pv: Balance) -> DispatchResult {
+		self.values
+			.try_push((id, pv))
+			.map_err(|_| DispatchError::Other("Max portfilio value reached"))?;
+
+		Ok(self.value.ensure_add_assign(pv)?)
+	}
+
+	pub fn update_elem(&mut self, id: ElemId, new_pv: Balance) -> Result<bool, DispatchError> {
+		let old_pv = self
+			.values
+			.iter_mut()
+			.find(|(elem_id, _)| *elem_id == id)
+			.map(|(_, value)| value)
+			.ok_or(DispatchError::CannotLookup)?;
```

Comment:
I think this should be another error.

Reply:
Do you mean a new defined error or a `DispatchError::Other("")`?

This is an error that can not happen (only if things are badly programmed), so it's not testable. I usually only use defined errors at pallet level errors that can happen from the user interaction, and then I can test & document them correctly. If not, I use `Other`, or `CannotLookup`.